controllers/appwrapper_controller.go

```diff
@@ -119,44 +117,84 @@ func (r *AppWrapperReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
 	//onDelete replacement
+	if !appwrapper.ObjectMeta.DeletionTimestamp.IsZero() {
+		if containsString(appwrapper.ObjectMeta.Finalizers, finalizerName) {
+			if err := r.finalizeScalingDownMachines(ctx, &appwrapper); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Remove the finalizer from the object's metadata
+			appwrapper.ObjectMeta.Finalizers = removeString(appwrapper.ObjectMeta.Finalizers, finalizerName)
+			if err := r.Update(ctx, &appwrapper); err != nil {
+				return ctrl.Result{}, err
+			}
+		} else {
+			// The object is being deleted but doesn't have our finalizer, so we add it
+			appwrapper.ObjectMeta.Finalizers = append(appwrapper.ObjectMeta.Finalizers, finalizerName)
```

Comment:
The finaliser should be added the first time we reconcile the appWrapper, or simply every time it's reconciled, checking the presence of the finaliser, and only when machineSet is used and re-used.